rclcpp/include/rclcpp/node_interfaces/node_base.hpp

```diff
@@ -145,4 +156,23 @@ class NodeBase : public NodeBaseInterface
 }  // namespace node_interfaces
 }  // namespace rclcpp
 
+// Class to hold the global map of mutexes
+class rclcpp::node_interfaces::map_of_mutexes
+{
+public:
+  map_of_mutexes();
+  ~map_of_mutexes();
+
+  // Methods need to be protected by internal mutex
+  void create_mutex_of_nodebase(const rclcpp::node_interfaces::NodeBaseInterface * nodebase);
+  std::shared_ptr<std::mutex>
+  get_mutex_of_nodebase(const rclcpp::node_interfaces::NodeBaseInterface * nodebase);
+  void delete_mutex_of_nodebase(const rclcpp::node_interfaces::NodeBaseInterface * nodebase);
+
+  // Members
+  std::unordered_map<const rclcpp::node_interfaces::NodeBaseInterface *,
+    std::shared_ptr<std::mutex>> data;
+  std::mutex internal_mutex;
+};
```

Comment:
I'll suggest just moving this class up above the `NodeBase` definition.  That way you can get rid of the forward-reference to it, and also just define it in the proper namespace.